### PR TITLE
`Lectures`: Add dismiss modal for unsaved changes to title or period section

### PR DESCRIPTION
**Pull request by florian-glombik**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
With [Lectures: Add status bar to lecture creation and edit mode #9655](https://github.com/ls1intum/Artemis/pull/9655) the guided mode will be removed and a "Status-bar-edit-view" will be introduced, with the sections `title`, `period`, `attachments`,  and `units`.

As `attachments` and `units` are saved directly when being edited, but for `title` and `period` a save button will explicitly need to be pressed the user shall not be able to leave the edit page mistakenly without saving the changes to the `title` and `period` section. 
To keep the scope of #9655 at a manageable size, we introduce the warning modal beforehand.

### Description
- Introduced signals in touched components
- Added a new modal that is displayed when the lecture edit view _(not in guided mode, as it will be removed soon)_ is closed without saving changes to the `title` or `period` section
- Added `isPeriodSectionValid` computed signal to period section component, also renamed and moved the component as the wizard will be removed soon
-  Removed duplicated datepicker code in `lecture-update.component.html` and using `lecture-period.component.ts` instead
- Disabled save button for lecture if no changes were made to title / period section
- Removed unused translations `competenciesStepTitle` and `competenciesStepMessage`
- Adjusted and extended client tests

### Steps for Testing
Prerequisites:
- 1 Instructor

1. Create a Lecture (in normal mode)
2. Verify all entries are saved properly (title, shortname, description, dates)
3. Edit the Lecture
4. Verify the edited information was saved properly
5. Verify that the `Save` button is only enabled if you made a change to the lecture
6. Edit the lecture and navigate somewhere else without saving
7. See that modal appears which asks if the changes shall really be dismissed
8. Verify that with cancel you stay on the edit page
9. Verify that with save the changes are dismissed

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| exercise-title-channel-name.component.ts | 96.77% |                           ✅  |
| close-edit-lecture-modal.component.ts | 88.88% |             (no significant logic in component)               ❌ |
| hasLectureUnsavedChanges.guard.ts | 100% |                            ✅ |
| lecture-period.component.ts | 60% |                         (no significant logic in component)   ❌ |
| lecture-title-channel-name.component.ts | 100% |                            ✅ |
| lecture-update.component.ts | 90.09% |                            ✅ |
| lecture.route.ts | 0% |                            ❌ |
| title-channel-name.component.ts | 100% |                            ✅ |

### Screenshots
saveButtonDisabledAndDismissModal-10023

image

## Release Notes

- **New Features**
  - Introduced a modal for handling unsaved changes when editing lectures.
  - Added a guard to prevent navigation away from the lecture update component if there are unsaved changes.
  - Added a new component for updating lecture periods with real-time validation of date inputs.
  - Enhanced the `TitleChannelNameComponent` and `ExerciseTitleChannelNameComponent` for streamlined output event handling.
  - Updated the lecture update wizard to simplify user interactions.

- **Bug Fixes**
  - Updated input and output handling in various components to ensure proper event emission and data binding.

- **Documentation**
  - Updated localization files to enhance user prompts regarding unsaved changes and removed outdated competency-related messages.

- **Tests**
  - Added new test specifications for the modal and guard functionalities, ensuring comprehensive coverage of their behaviors.

### CODE REVIEW

File: src/main/webapp/app/lecture/lecture-update.component.ts [hunk 1]
```diff
@@ -82,6 +118,42 @@ export class LectureUpdateComponent implements OnInit {
...
+
+        this.lectureOnInit = cloneDeep(this.lecture());
+    }
+
+    ngOnDestroy() {
+        this.subscriptions.unsubscribe();
+    }
+
+    isChangeMadeToTitleSection() {
+        return (
```
Comment (ole-ve) [inline src/main/webapp/app/lecture/lecture-update.component.ts:130]: Really a minor but for all these comparisons, you could consider using some functionality of `loadash` like `pick` and `isEqual`.

File: src/test/javascript/spec/component/lecture/lecture-update.component.spec.ts [hunk 2]
```diff
@@ -45,18 +51,22 @@ describe('LectureUpdateComponent', () => {
         pastLecture.endDate = yesterday;
 
         TestBed.configureTestingModule({
-            imports: [ArtemisTestModule, FormsModule, MockModule(NgbTooltipModule)],
+            imports: [ArtemisTestModule, ArtemisSharedModule, FormsModule, MockModule(NgbTooltipModule), MockModule(OwlDateTimeModule)],
             declarations: [
                 LectureUpdateComponent,
-                MockComponent(LectureTitleChannelNameComponent),
+                LectureTitleChannelNameComponent,
```
Comment (ole-ve) [inline src/test/javascript/spec/component/lecture/lecture-update.component.spec.ts:57]: Can we mock some of those or are all required within the tests?

**Review by ole-ve — DISMISSED** [review]
code looks reasonable, added two smaller ideas

File: src/test/javascript/spec/component/lecture/lecture-update.component.spec.ts (hunk 2 above)
Comment (florian-glombik) [reply src/test/javascript/spec/component/lecture/lecture-update.component.spec.ts:57]: We do not seem to need the ArtemisSharedModule import here, besides that I am not aware that we could optimize iz further.
I would assume that ngMocks still have problems with mocking signals, but I did not investigate it in this particular case.

File: src/main/webapp/app/lecture/lecture-update.component.ts (hunk 1 above)
Comment (florian-glombik) [reply src/main/webapp/app/lecture/lecture-update.component.ts:130]: I tried it, but this does not handle edge cases properly, so I think the current implementation is the easiest to understand 

When using this approach
```
    isChangeMadeToTitleSection() {
        const current = pick(this.lecture(), ['title', 'channelName', 'description']);
        const initial = pick(this.lectureOnInit, ['title', 'channelName', 'description']);
        return !isEqual(current, initial);
    }
```

The following test fails

```
        it('should handle undefined from description properly', () => {
            lectureUpdateComponent.lecture.set({ title: 'new title', channelName: 'new channel', description: 'new description' } as Lecture);
            lectureUpdateComponent.lectureOnInit = { title: 'old title', channelName: 'old channel', description: undefined } as Lecture;
            expect(lectureUpdateComponent.isChangeMadeToTitleSection()).toBeTrue();

            lectureUpdateComponent.lecture.set({
                title: lectureUpdateComponent.lectureOnInit.title,
                channelName: lectureUpdateComponent.lectureOnInit.channelName,
                description: '', // will be an empty string if the user clears the input, but was loaded with undefined in that case
            } as Lecture);
            expect(lectureUpdateComponent.isChangeMadeToTitleSection()).toBeFalse();
        });
```

**Review by az108 — APPROVED** [review]
Code LGTM

**Review by ole-ve — DISMISSED** [review]

**Review by ole-ve — APPROVED** [review]
re-approve after changing translations

**Review by az108 — APPROVED** [review]
reapprove code after changes 👍

**Review by rabeatwork — APPROVED** [review]

**Review by sachmii — APPROVED** [review]
Tested on TS2, everything works perfectly.

**Pull request merged**